File: 020_ValidParentheses.cpp

//A solution from LeetCode
class Solution {
public:
    bool isValid(string s) {
        stack<char> paren;
        for (char& c : s) {
            switch (c) {
                case '(': 
                case '{': 
                case '[': paren.push(c); break;
                case ')': if (paren.empty() || paren.top()!='(') return false; else paren.pop(); break;
                case '}': if (paren.empty() || paren.top()!='{') return false; else paren.pop(); break;
                case ']': if (paren.empty() || paren.top()!='[') return false; else paren.pop(); break;
                default: ; // pass
            }
        }
        return paren.empty() ;
    }
};
//22.85%
class Solution {
public:
    bool isValid(string s) {
        stack<char> charStack;
		for(string::iterator iteS = s.begin(); iteS != s.end(); iteS++)
		{
			switch(*iteS)
			{
				case '(':
				case '[':
				case '{':
					charStack.push(*iteS);
					break;
				case ')':
					if( charStack.empty() || charStack.top() != '(')
						return false;
					charStack.pop();
					break;
				case ']':
					if( charStack.empty() || charStack.top() != '[')
						return false;
					charStack.pop();
					break;
				case '}':
					if( charStack.empty() || charStack.top() != '{')
						return false;
					charStack.pop();
					break;
				default:
					return false;
			}
		}
		
		return charStack.empty();
    }
};

File: 020_ValidParentheses.py

#Solution from LeetCode
class Solution:
    # @return a boolean
    def isValid(self, s):
        stack = []
        dict = {"]":"[", "}":"{", ")":"("}
        for char in s:
            if char in dict.values():
                stack.append(char)
            elif char in dict.keys():
                if stack == [] or dict[char] != stack.pop():
                    return False
            else:
                return False
        return stack == []

#45.4%
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        strMap = {}
        strMap[')'] = '('
        strMap['}'] = '{'
        strMap[']'] = '['
        
        valueList = ['(', '{', '[']
        keyList = [')', '}', ']']
        strList = []
        for str in s:
            if str in keyList:
                if len(strList) == 0 or strMap[str] != strList.pop():
                    return False
            elif str in valueList:
                strList.append(str)
            else:
                return False
        
        if len(strList) == 0:
            return True
        return False